### DISCUSSION TITLE
Deconstruct for interface

### DISCUSSION THREAD

**Discussion by WeihanLi** [dotnet/csharplang#6500] [General] [answered]
Is it possible to use deconstruct for an interface?

For example, we could use `deconstruct` for Dictionary, is it possible to use deconstruct for `IDictionary`?

``` c#
var dic = new Dictionary<string, string>();
foreach (var (key, value) in dic)

```

**Comment by HaloFour** [comment, marked as answer]
The short answer is yes, any type can support custom deconstruction via extension methods.

However, if you're referring to the non-generic `System.Collections.IDictionary` this is complicated by the fact that it implements `System.Collections.IEnumerable` which effectively enumerates all values as `System.Object`.  An extension method to deconstruct `System.Object` is possible but wouldn't apply in most situations, so what I might suggest is an extension method on `IDictionary` that returns `IEnumerable<DictionaryEntry>` which would be safe to deconstruct:

```cs
public static class IDictionaryExtensions {
    // returns a type-safe enumerable of each entry
    public static IEnumerable<DictionaryEntry> Entries(this IDictionary dictionary) {
        foreach (DictionaryEntry entry in dictionary) {
            yield return entry;
        }
    }
    
    // deconstruct each entry
    public static void Deconstruct(this DictionaryEntry entry, out object key, out object value) {
        key = entry.Key;
        value = entry.Value;
    }
}
```

Then you can use it like this:

```cs
IDictionary dictionary = ...;

foreach (var (key, value) in dictionary) {
    // ...
}
```

However, the key and value will both be of type `System.Object`.
